File: src/eQuantic.UI.Components/Badge.cs
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Components;

/// <summary>
/// The design system's Badge (spec B7): a non-interactive count/dot annotation. Dot 8dp; count pill
/// 16dp min height, Radius.Full, 10/700 label, 4dp X padding, "99+" past <see cref="Max"/>.
/// Destructive pair = attention (default), neutral SurfaceSubtle = passive counts, status pairs
/// allowed. v1 renders INLINE — the −4/−4 overlay attach composes via Stack + Positioned when the
/// Stack primitive lands; the 2dp Surface ring applies only to that overlay case.
/// </summary>
public sealed class Badge : StatelessComponent
{
    public Badge(int count = 0, int max = 99, Variant variant = Variant.Destructive)
    {
        Count = count;
        Max = max;
        Variant = variant;
    }

    public int Count { get; init; }
    public int Max { get; init; }
    public Variant Variant { get; init; }

    /// <summary>Neutral passive-count styling (SurfaceSubtle pair) instead of a variant pair.</summary>
    public bool Neutral { get; init; }

    /// <summary>The 8dp presence dot — no count.</summary>
    public bool Dot { get; init; }

    /// <summary>The 2dp Surface ring the spec requires when the badge overlays content
    /// (attach via Stack + Positioned(top: −4, end: −4)).</summary>
    public bool Ring { get; init; }

    public static Badge AsDot(Variant variant = Variant.Destructive) => new(0, 99, variant) { Dot = true };

    public override VisualNode Build(ComponentContext context)
    {
        var theme = context.Theme;
        var fill = Neutral ? theme.SurfaceSubtle : theme.Colors(Variant).Base;
        var textColor = Neutral ? theme.TextSecondary : theme.Colors(Variant).OnBase;

        if (Dot)
        {
            return new Box(new BoxStyle
            {
                Width = Ring ? 12 : 8,
                Height = Ring ? 12 : 8,
                Background = fill,
                CornerRadius = new CornerRadii(theme.Shape(ShapeScale.Full)),
                BorderWidth = Ring ? 2f : 0f,
                BorderColor = theme.Surface,
            });
        }

        var label = Count > Max ? $"{Max}+" : $"{Count}";
        var text = new Text(label, TypeRole.Caption, textColor, maxLines: 1)
        {
            StyleOverride = new TypeStyle(10, 12, FontWeight.Bold, 0, 1.3f),
        };

        var content = new Row(gap: 0) { Main = MainAlign.Center, Height = SizeValue.Fill };
        content.Add(text);

        return new Box(new BoxStyle
        {
            Height = Ring ? 20 : 16,
            MinWidth = Ring ? 20 : 16,
            Padding = EdgeInsets.Symmetric(Space.S1, 0),
            Background = fill,
            CornerRadius = new CornerRadii(theme.Shape(ShapeScale.Full)),
            BorderWidth = Ring ? 2f : 0f,
            BorderColor = theme.Surface,
        }, content);
    }
}

File: src/eQuantic.UI.Runtime/src/shared/__transpiled__/Badge.ts
import { Box, BoxStyle, BuildContext, CornerRadii, EdgeInsets, Row, SizeValue, Space, StatelessComponent, Text, TypeStyle } from "@equantic/runtime";

export class Badge extends StatelessComponent {
    declare count: number;
    declare max: number;
    declare variant: string;
    declare neutral: boolean;
    declare dot: boolean;
    declare ring: boolean;
    constructor(count: any = 0, max: any = 99, variant: any = 'destructive', props?: any) {
        super(props);
        if (count !== undefined) this.count = count;
        if (max !== undefined) this.max = max;
        if (variant !== undefined) this.variant = variant;
        this.count = count;this.max = max;this.variant = variant;
    }

    build(context: BuildContext) {
        let theme = context.theme;let fill = this.neutral ? theme.surfaceSubtle : theme.colors(this.variant).base;let textColor = this.neutral ? theme.textSecondary : theme.colors(this.variant).onBase;if (this.dot) {return new Box(new BoxStyle({ width: this.ring ? 12 : 8, height: this.ring ? 12 : 8, background: fill, cornerRadius: new CornerRadii(theme.shape('full')), borderWidth: this.ring ? 2 : 0, borderColor: theme.surface }));}let label = this.count > this.max ? `${this.max}+` : `${this.count}`;let text = new Text(label, 'caption', textColor, 1, { styleOverride: new TypeStyle(10, 12, 'bold', 0, 1.3) });let content = new Row(0, { main: 'center', height: SizeValue.fill });content.add(text);return new Box(new BoxStyle({ height: this.ring ? 20 : 16, minWidth: this.ring ? 20 : 16, padding: EdgeInsets.symmetric(Space.s1, 0), background: fill, cornerRadius: new CornerRadii(theme.shape('full')), borderWidth: this.ring ? 2 : 0, borderColor: theme.surface }), content);
    }

    static asDot(variant: Variant) {
        return { dot: true };
    }

}

